qualtran/drawing/qpic_diagram.py

```diff
@@ -0,0 +1,332 @@
+            return f'{wire}' if symbol.filled else f'-{wire}', None
+        elif isinstance(symbol, ModPlus):
+            return f'+{wire}', None
+        symbol_text = _soq_symbol_to_text(symbol)
+        width = _soq_symbol_width(symbol)
+        return f'{wire} ', f'G:width={width} {symbol_text}'
+
+    def _dealloc_wire_for_soq(self, soq: Soquet) -> None:
+        self._wire_manager.dealloc_wire_for_soq(self.soq_map[soq])
+        self.soq_map.pop(soq)
+
+    def _alloc_wire_for_soq(self, soq: Soquet) -> None:
+        self.soq_map[soq] = soq
+        wire_name = self._wire_manager.alloc_wire_for_soq(soq)
+        if wire_name in self.allocated_wires:
+            return
+        self.allocated_wires.add(wire_name)
+        # For a soquet on THRU/LEFT register, a wire should already be allocated.
+        # assert soq.reg.side == Side.RIGHT
+        if isinstance(soq.binst, DanglingT):
+            # A RIGHT wire on a LeftDangle bloq should start at the beginning.
+            assert soq.binst == LeftDangle
+            wire_label = self._wire_manager.soq_to_wirelabel(soq)
+            self.wires += [f'{wire_name} W {wire_label}']
+            if soq.reg.dtype != QBit():
+                dtype_str = _format_label_text(str(soq.reg.dtype), scale=0.5)
+                self.wires += [f'{wire_name} / {dtype_str}']
+
+        else:
+            # A RIGHT wire on an intermediate Bloq should start as off and be turned on
+            # by the Bloq.
+            self.wires += [f'{wire_name} W off']
+
+    def add_connection(self, cxn: 'Connection'):
+        if cxn.left not in self.soq_map:
+            self._alloc_wire_for_soq(cxn.left)
+        self.soq_map[cxn.right] = self.soq_map[cxn.left]
+
+    def add_bloq(self, bloq: 'Bloq', pred: List['Connection'], succ: List['Connection']) -> None:
+        controls, targets = [], []
+
+        def add_soq(soq: Soquet):
+            wire, gate = self._add_soq(soq)
+            if gate is None:
+                controls.append(wire)
+            else:
+                targets.append(wire + gate)
+
+        for cxn in pred:
+            self.add_connection(cxn)
+            add_soq(cxn.right)
+
+        for cxn in succ:
+            self.add_connection(cxn)
+            if cxn.left.reg.side == Side.RIGHT:
+                add_soq(cxn.left)
+        self.gates += [' '.join(targets + controls)]
+
+
+def get_qpic_data(bloq: 'Bloq', file_path: Union[None, pathlib.Path, str] = None) -> List[str]:
+    """Get the input data that can be used to draw a latex diagram for `bloq` using `qpic`.
+
+    Args:
+        bloq: Bloqs to draw the latex diagram for
+        file_path: If specified, the output is stored at the file.
+
+    Returns:
+        A list of strings representing the input to `qpic` utility to draw a latex diagram
+        for bloq. If specified, the output is written to `file_path`.
+    """
+    cbloq = bloq.as_composite_bloq()
+    qpic_circuit = QpicCircuit(cbloq.signature)
```

Comment:
What do you think about removing all the business logic from the QpicCircuit constructor? i.e.: in `__init__` just initialize the member variables; then in this function you can call

```python
qpic_circuit.set_up_signature(cbloq.signature)
for binst, pred, succ in cbloq.iter_bloqnections():
  qpic_circuit.add_bloq(pred, succ)
```

Reply:
done